### PR TITLE
[SYCL-MLIR]: constructAttributeList returns CallingConv

### PR DESCRIPTION
**Pull request by etiotto**
This PR modifies `constructAttributeList` to pass CallingConv as a reference argument (same as clang). The PR also applies a few clang-tidy related changes.

### CODE REVIEW

File: polygeist/tools/cgeist/Lib/CGCall.cc [hunk 1]
```diff
@@ -100,21 +100,21 @@ static void castCallerArgs(mlir::func::FuncOp Callee,
...
     ArrayRef<std::pair<ValueCategory, clang::Expr *>> Arguments,
     QualType RetType, bool RetReference, clang::Expr *Expr,
-    const FunctionDecl *Callee) {
+    const FunctionDecl &Callee) {
   SmallVector<mlir::Value, 4> Args;
-  auto FnType = Tocall.getFunctionType();
-  const clang::CodeGen::CGFunctionInfo &FI =
-      Glob.GetOrCreateCGFunctionInfo(Callee);
-  auto FIArgs = FI.arguments();
+  mlir::FunctionType FnType = ToCall.getFunctionType();
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/CGCall.cc:108]: Just NFC changes.

File: polygeist/tools/cgeist/Lib/CGCall.cc [hunk 2]
```diff
@@ -412,13 +413,12 @@ ValueCategory MLIRScanner::callHelper(
...
   }
-  if (Op->getNumResults()) {
+
+  if (Op->getNumResults())
     return ValueCategory(Op->getResult(0),
                          /*isReference*/ RetReference);
-  }
+
   return nullptr;
-  llvm::errs() << "do not support indirecto call of " << Tocall << "\n";
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/CGCall.cc:420]: Dead code

File: polygeist/tools/cgeist/Lib/CGCall.cc [hunk 3]
```diff
@@ -1149,17 +1133,16 @@ ValueCategory MLIRScanner::VisitCallExpr(clang::CallExpr *Expr) {
   if (auto *CC = dyn_cast<CXXMemberCallExpr>(Expr)) {
     ValueCategory Obj = Visit(CC->getImplicitObjectArgument());
     ObjType = CC->getObjectType();
-#ifdef DEBUG
-    if (!obj.val) {
+    LLVM_DEBUG(if (!Obj.val) {
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/CGCall.cc:1136]: NFC changes

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 4]
```diff
@@ -862,19 +862,20 @@ ValueCategory MLIRScanner::VisitConstructCommon(clang::CXXConstructExpr *cons,
     ShouldEmit = true;
 
   FunctionToEmit F(*ctorDecl, mlirclang::getInputContext(builder));
-  auto tocall = cast<func::FuncOp>(Glob.GetOrCreateMLIRFunction(F, ShouldEmit));
+  auto ToCall = cast<func::FuncOp>(Glob.GetOrCreateMLIRFunction(F, ShouldEmit));
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:865]: NFC changes

File: polygeist/tools/cgeist/Lib/CodeGenTypes.cc [hunk 5]
```diff
@@ -646,16 +646,12 @@ CodeGenTypes::getFunctionType(const clang::CodeGen::CGFunctionInfo &FI,
 void CodeGenTypes::constructAttributeList(
     StringRef Name, const clang::CodeGen::CGFunctionInfo &FI,
     clang::CodeGen::CGCalleeInfo CalleeInfo, mlirclang::AttributeList &AttrList,
-    bool AttrOnCallSite, bool IsThunk) {
+    unsigned &CallingConv, bool AttrOnCallSite, bool IsThunk) {
   MLIRContext *Ctx = TheModule->getContext();
   mlirclang::AttrBuilder FuncAttrsBuilder(*Ctx);
   mlirclang::AttrBuilder RetAttrsBuilder(*Ctx);
 
-  unsigned CC = FI.getEffectiveCallingConvention();
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/CodeGenTypes.cc:654]: Caller will set calling convention (same as clang).

File: polygeist/tools/cgeist/Lib/clang-mlir.cc [hunk 6]
```diff
@@ -3044,30 +3043,37 @@ void MLIRASTConsumer::setMLIRFunctionAttributes(
...
+                                      /*AttrOnCallSite*/ false,
                                       /*IsThunk*/ false);
 
     // Set additional function attributes that are not derivable from the
     // function declaration.
-    mlirclang::AttrBuilder attrBuilder(*ctx);
+    mlirclang::AttrBuilder attrBuilder(*Ctx);
     {
-      LLVM::Linkage lnk = getMLIRLinkage(getLLVMLinkageType(FD, ShouldEmit));
+      attrBuilder.addAttribute(
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/clang-mlir.cc:3060]: Used to be set in `constructAttributeList`, that function should only produce information (not change the IR).

File: polygeist/tools/cgeist/Lib/CGCall.cc (hunk 3 above)
Comment (whitneywhtsang) [reply polygeist/tools/cgeist/Lib/CGCall.cc:1136]: Please add `{}` inside `LLVM_DEBUG()`.

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 7]
```diff
@@ -862,19 +862,20 @@ ValueCategory MLIRScanner::VisitConstructCommon(clang::CXXConstructExpr *cons,
...
 
-  SmallVector<std::pair<ValueCategory, clang::Expr *>> args;
-  args.emplace_back(std::make_pair(obj, (clang::Expr *)nullptr));
-  for (auto a : cons->arguments())
-    args.push_back(std::make_pair(Visit(a), a));
-  callHelper(tocall, innerType, args,
+  SmallVector<std::pair<ValueCategory, clang::Expr *>> Args;
+  Args.emplace_back(std::make_pair(obj, (clang::Expr *)nullptr));
+  for (auto A : cons->arguments())
+    Args.push_back(std::make_pair(Visit(A), A));
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:870]:
```suggestion
  SmallVector<std::pair<ValueCategory, clang::Expr *>> Args{{obj, nullptr}};
  Args.reserve(cons->getNumArgs() + 1);
  for (auto A : cons->arguments())
    Args.emplace_back(Visit(A), A);
```

**Review by victor-eds — CHANGES_REQUESTED** [review]

File: polygeist/tools/cgeist/Lib/CGCall.cc (hunk 3 above)
Comment (etiotto) [reply polygeist/tools/cgeist/Lib/CGCall.cc:1136]: I like the way it looks now.

Comment (etiotto) [reply polygeist/tools/cgeist/Lib/CGCall.cc]: done

**Review by whitneywhtsang — APPROVED** [review]

**Review by victor-eds — APPROVED** [review]

**Comment by etiotto** [comment]
@whitneywhtsang has added the parallel_for.cpp test to the LLVM test suite, that test is failing but is not a new failure. PR #7342 should make that test pass. Committing this one anyway because it is not breaking an existing working test.

**Pull request merged**